src/api/procedures/utils.ts

```diff
@@ -322,3 +330,166 @@ export function assertRequirementsNotTooComplex(
+    case AuthorizationType.BecomeAgent:
+      // no additional checks
+      return;
+    case AuthorizationType.AddMultiSigSigner:
+      // no additional checks
+      return;
+    case AuthorizationType.PortfolioCustody:
+      // no additional checks
+      return;
+    case AuthorizationType.JoinIdentity:
+      return assertJoinIdentityAuthorizationValid(authRequest);
+    case AuthorizationType.AddRelayerPayingKey:
+      return assertAddRelayerPayingKeyAuthorizationValid(authRequest.data);
+    default:
+      throw new UnreachableCaseError(authRequest.data); // ensures switch statement covers all values
+  }
+}
+
+/**
+ * asserts valid  primary key rotation authorization
+ */
+export async function assertPrimaryKeyRotationAuthorizationValid(
+  authRequest: AuthorizationRequest
+): Promise<void> {
+  if (authRequest.target instanceof Identity) {
+    throw new PolymeshError({
+      code: ErrorCode.ValidationError,
+      message: 'An Identity can not join another Identity',
+    });
+  }
+}
+
+/**
+ * asserts valid attest primary key authorization
+ */
+export async function assertAttestPrimaryKeyAuthorizationValid(
+  authRequest: AuthorizationRequest
+): Promise<void> {
+  if (!(await authRequest.issuer.isCddProvider())) {
+    throw new PolymeshError({
+      code: ErrorCode.UnmetPrerequisite,
+      message: 'Issuer must be a CDD provider',
+    });
+  }
+}
+
+/** asserts transfer ticker authorization is valid */
+export async function assertTransferTickerAuthorizationValid(
+  context: Context,
+  data: GenericAuthorizationData
+): Promise<void> {
+  const reservation = new TickerReservation({ ticker: data.value }, context);
+  const { status } = await reservation.details();
+  if (status === TickerReservationStatus.Free) {
+    throw new PolymeshError({
+      code: ErrorCode.UnmetPrerequisite,
+      message: 'The Ticker is not reserved',
+    });
+  }
+  if (status === TickerReservationStatus.TokenCreated) {
+    throw new PolymeshError({
+      code: ErrorCode.UnmetPrerequisite,
+      message: 'The ticker has already been used to create an Asset',
+    });
+  }
+}
+
+/** asserts valid transfer asset ownership authorization */
+export async function assertTransferAssetOwnershipAuthorizationValid(
+  context: Context,
+  data: GenericAuthorizationData
+): Promise<void> {
+  const token = new SecurityToken({ ticker: data.value }, context);
+  if (!(await token.exists()))
+    throw new PolymeshError({
+      code: ErrorCode.UnmetPrerequisite,
+      message: 'The Asset does not exist',
+    });
+}
+
+/**
+ * asserts valid join identity authorization request
+ */
+export async function assertJoinIdentityAuthorizationValid(
+  authRequest: AuthorizationRequest
+): Promise<void> {
+  // https://github.com/PolymathNetwork/Polymesh/blob/5fec16fdfb05dc9022508880503e835ae9c1776c/pallets/identity/src/keys.rs#L470
+  const identity = authRequest.issuer;
+  if (!(await identity.hasValidCdd())) {
```

Comment:
We avoid inline awaits for better readability (there are quite a few in this PR)